source/extensions/transport_sockets/starttls/starttls_socket.h

```diff
@@ -56,6 +57,41 @@ class StartTlsSocket : public Network::TransportSocket, Logger::Loggable<Logger:
   }
 
 private:
+  // This is a proxy for wrapping the transport callback object passed from the consumer.
+  // Its primary purpose is to filter Connected events to ensure they only happen once per open.
+  // connection open.
+  class CallbackProxy : public Network::TransportSocketCallbacks {
+  public:
+    CallbackProxy(Network::TransportSocketCallbacks* callbacks) : parent_(callbacks) {}
+
+    Network::IoHandle& ioHandle() override { return parent_->ioHandle(); }
+    const Network::IoHandle& ioHandle() const override {
+      return const_cast<const Network::TransportSocketCallbacks*>(parent_)->ioHandle();
```

Comment:
You shouldn't need `const_cast` to add const.

Reply:
This is not to return `const` value, but to add `const` to `parent_`, so the compiler will select `const` version of `ioHandle`.
 https://github.com/envoyproxy/envoy/blob/main/envoy/network/transport_socket.h#L51-L56
Without the `const`, the compiler invokes `non-const` version of `ioHandle`.

The other way of doing it would be `static_cast` or `const` local variable.

Reply:
I know what you're trying to do, but `const_cast` is to remove but not add. You may want to use `std::as_const`.

Reply:
Yes, you are right. Thanks. Changed to `static_cast.`